main/src/ca/uwaterloo/flix/language/phase/Parser2.scala

```diff
@@ -1856,18 +1869,28 @@ object Parser2 {
       close(mark, TreeKind.Expr.Unary)
     }
 
+    /**
+      * The condition and then sub-trees will always exist, but the else sub-tree is optional.
+      * If condition/then branch is missing in the source code, it will be represented as an Expr.Error inside that sub-tree.
+      */
     private def ifThenElseExpr()(implicit s: State): Mark.Closed = {
       assert(at(TokenKind.KeywordIf))
       val mark = open()
       expect(TokenKind.KeywordIf, SyntacticContext.Expr.OtherExpr)
       expect(TokenKind.ParenL, SyntacticContext.Expr.OtherExpr)
-      expression()
+      val condMark = open()
+      expression(exitTokens = Set(TokenKind.ParenR))
+      close(condMark, TreeKind.Expr.Condition)
       expect(TokenKind.ParenR, SyntacticContext.Expr.OtherExpr)
-      expression()
+      val thenMark = open()
+      expression(exitTokens = Set(TokenKind.KeywordElse))
+      close(thenMark, TreeKind.Expr.Then)
       if (eat(TokenKind.KeywordElse)) {
         // Only call expression, if we found an 'else'. Otherwise when it is missing, defs might
         // get read as let-rec-defs.
+        val elseMark = open()
         expression()
```

Comment:
This one can have exit token `Else` to better parse `if (true) 1 + else 2` I believe